# Primal, Dual and Complementary Slackness

In [19]:
import numpy as np
import sympy
from sympy import *
from scipy.optimize import linprog

import warnings
warnings.filterwarnings('ignore')

# Optimization Problem

*PAR* is a small manufacturer of golf equipment and accessories whose distributor convinced them that there is a market for both the mid-priced golf bag, known as the standard model, and for a high-priced golf bag, known as the deluxe model. The distributor is so confident in the market that if *PAR* can manufacture the bags competitively, the distributor agrees to purchase all the bags *PAR* can produce over the next three months.

A careful analysis of the manufacturing requirements resulted in the following table, which shows the time and production needs for the four required manufacturing operations and the accounting department's estimate of the contribution to profit per bag:

| **Product** 	| **Cutting and Dyeing (h)** 	| **Sewing (h)** 	| **Finishing (h)** 	| **Inspection and Packaging (h)** 	| **Profit per bag ($)** 	|
|--------------	|-----------------------------	|-----------------	|-------------------	|----------------------------------	|-------------------------	|
| Standard     	| 7/10                        	| 1/2             	| 1                 	| 1/10                             	| 7                       	|
| Deluxe       	| 1                           	| 5/6             	| 2/3               	| 1/4                              	| 9                       	|

The manufacturing director estimates that during the next three months, there will be 630 hours of cutting and dyeing time, 600 hours of sewing time, 708 hours of finishing time, and 135 hours of inspection and packaging time available for the production of golf bags.

#### If *PAR* wishes to maximize its profit, what amount of standard and deluxe bags should it manufacture?


# Problem 1

#### Solve the problem using the primal and dual form of the optimization problem.

#### Original Optimization Problem Statement

Decision Variables:

$x_1 \longrightarrow $ Quantity of standard bags

$x_2 \longrightarrow $ Quantity of deluxe bags

Objective Function: Profit

$f(x) = 7x_1 + 9x_2$

Constraints: Time in hours

$\frac{7}{10}x_1 + x_2 \leq 630 $ 

$\frac{1}{2}x_1 + \frac{5}{6}x_2 \leq 600$ 

$x_1 + \frac{2}{3}x_2 \leq 708$

$\frac{1}{10}x_1 + \frac{1}{4}x_2 \leq 135$

Original Optimization Problem: 

max $f(x) = 7x_1 + 9x_2$ s.t.

$\frac{7}{10}x_1 + x_2 \leq 630 $ 

$\frac{1}{2}x_1 + \frac{5}{6}x_2 \leq 600$ 

$x_1 + \frac{2}{3}x_2 \leq 708$ 

$\frac{1}{10}x_1 + \frac{1}{4}x_2 \leq 135$

Primal Problem: 

min $-f(x) = -7x_1 - 9x_2$ s.t.

$-\frac{7}{10}x_1 - x_2 \geq -630 $ 

$-\frac{1}{2}x_1 - \frac{5}{6}x_2 \geq -600$ 

$-x_1 - \frac{2}{3}x_2 \geq -708$ 

$-\frac{1}{10}x_1 - \frac{1}{4}x_2 \geq -135$

Dual Problem:

max $q(\lambda) = -630\lambda_1 -600\lambda_2 - 708\lambda_3 - 135\lambda_4$ s.t. 

$-\frac{7}{10}\lambda_1 - \frac{1}{2}\lambda_2 - \lambda_3 - \frac{1}{10}\lambda_4 \leq -7$ 

$-\lambda_1 - \frac{5}{6}\lambda_2 - \frac{2}{3}\lambda_3 - \frac{1}{4}\lambda_4 \leq -9 $


In [20]:
c = np.array([[-7, -9]])
A = np.array([[7/10, 1],[1/2, 5/6],[1, 2/3],[1/10, 1/4]])
b = np.array([[630],[600],[708],[135]])

x1_bounds = (0, None)
x2_bounds = (0, None)


res = linprog(c, A_ub=A, b_ub=b, bounds = np.array([x1_bounds, x2_bounds]), method='simplex', options={"disp": False})
print(f'The solution to the original optimization problem is x = {res.x}')
print(f'The value of the objective function evaluated at the optimal value is {-1*res.fun}')


The solution to the original optimization problem is x = [540. 252.]
The value of the objective function evaluated at the optimal value is 6048.0


According to this result, it is obtained that **540 standard bags and 252 deluxe bags** should be manufactured to maximize the profit to a value of **$6,048**.

The profit obtained by manufacturing **only standard bags** corresponds to **$4,956**.

The profit obtained by manufacturing **only deluxe bags** corresponds to **$4,860**.

# Problem 2

### Different strategies

*PAR*'s store manager has proposed three different strategies to try and maximize profits. 

**Strategy 1:** Increase the price of deluxe bags such that its profit per bag is now $18

**Strategy 2:** Increase the price of standard bags such that the its per bag is now $14

**Strategy 3:** Increase the maximum amount of hours spent on cutting and dyeing to 1071, the maximum amount of hours spent on sewing to 1020, the maximum amount of hours of finishing time to 1203.6, and the maximum amount of hours spent on inspection and packaging  to 229.5.

Formulate an optimization problem for each strategy and decide which one is the best alternative to maximize profits for the company.

### a)
#### Formulation of the first strategy

Decision variables:

$x_1 \longrightarrow $ Quantity of standard bags \
$x_2 \longrightarrow $ Quantity of deluxe bags

Objective function: Profit

$f(x) = 7x_1 + 18x_2$

Constraints: Time in hours

$\frac{7}{10}x_1 + x_2 \leq 630 $ \
$\frac{1}{2}x_1 + \frac{5}{6}x_2 \leq 600$ \
$x_1 + \frac{2}{3}x_2 \leq 708$ \
$\frac{1}{10}x_1 + \frac{1}{4}x_2 \leq 135$

Optimization problem: \
max $f(x) = 7x_1 + 18x_2$ s.t.

$\frac{7}{10}x_1 + x_2 \leq 630 $ \
$\frac{1}{2}x_1 + \frac{5}{6}x_2 \leq 600$ \
$x_1 + \frac{2}{3}x_2 \leq 708$ \
$\frac{1}{10}x_1 + \frac{1}{4}x_2 \leq 135$

Primal problem: \
min $-f(x) = -7x_1 - 18x_2$ s.t.

$-\frac{7}{10}x_1 - x_2 \geq -630 $ \
$-\frac{1}{2}x_1 - \frac{5}{6}x_2 \geq -600$ \
$-x_1 - \frac{2}{3}x_2 \geq -708$ \
$-\frac{1}{10}x_1 - \frac{1}{4}x_2 \geq -135$


#### Formulation of the dual problem for the first strategy

max $q(\lambda) = 630\lambda_1 + 600\lambda_2 + 708\lambda_3 + 135\lambda_4$ s.t. \
$\frac{7}{10}\lambda_1 + \frac{1}{2}\lambda_2 + \lambda_3 + \frac{1}{10}\lambda_4 \geq -7$ \
$\lambda_1 + \frac{5}{6}\lambda_2 + \frac{2}{3}\lambda_3 + \frac{1}{4}\lambda_4 \geq -18 $


#### Solution of the dual problem for the first strategy

In [21]:
c = np.array([[630, 600, 708, 135]])
A = np.array([[-7/10, -1/2, -1, -1/10],[-1, -5/6, -2/3, -1/4],[-1,0,0,0],[0,-1,0,0],[0,0,-1,0],[0,0,0,-1]])
b = np.array([[-7],[-18],[0],[0],[0],[0]])

x1_bounds = (0, None)
x2_bounds = (0, None)
x3_bounds = (0, None)
x4_bounds = (0, None)


res = linprog(c, A_ub=A, b_ub=b, bounds = np.array([x1_bounds, x2_bounds, x3_bounds, x4_bounds]), method='simplex', options={"disp": True})
print(f'The solution to this optimization problem is λ = {res.x}')
print(f'The value of the objective function of the dual evaluated at the optimal value is {res.fun}')

Optimization terminated successfully.
         Current function value: 9720.000000 
         Iterations: 3
The solution to this optimization problem is λ = [ 0.  0.  0. 72.]
The value of the objective function of the dual evaluated at the optimal value is 9720.0


#### Solution of the primal via complementary slackness

From the previous solution, the complementary slackness conditions are evaluated to find the solution in the primal problem, i.e., the quantities of standard and deluxe bags to manufacture. The following constraints are obtained:

$-\frac{7}{10}(0) - \frac{1}{2}(0) - 1(0) -\frac{1}{10}(72) < -7$\
$-7.2 < -7 \longrightarrow$ Inactive constraint 

$-1(0) - \frac{5}{6}(0) - \frac{2}{3}(0) -\frac{1}{4}(72) < -18$
$-18 = -18 \longrightarrow$ Active constraint 

$\lambda_1^*=0\longrightarrow$ Active constraint 

$\lambda_2^*=0\longrightarrow$ Active constraint 

$\lambda_3^*=0\longrightarrow$ Active constraint 

$\lambda_4^*>0\longrightarrow$ Inactive constraint 

From the first and sixth constraints, it is obtained respectively that:

$x_1^*=0$ \
$-\frac{1}{10}x_1^*-\frac{1}{4}x_2^*=-135$

Finally, the system is solved and it is obtained that 0 standard bags and 540 deluxe bags should be manufactured.


### b) 
#### Formulation of the second strategy

Decision variables:

$x_1 \longrightarrow $ Quantity of standard bags \
$x_2 \longrightarrow $ Quantity of deluxe bags

Objective function: Profit

$f(x) = 14x_1 + 9x_2$

Constraints: Time in hours

$\frac{7}{10}x_1 + x_2 \leq 630 $ \
$\frac{1}{2}x_1 + \frac{5}{6}x_2 \leq 600$ \
$x_1 + \frac{2}{3}x_2 \leq 708$ \
$\frac{1}{10}x_1 + \frac{1}{4}x_2 \leq 135$

Optimization problem: \
max $f(x) = 14x_1 + 9x_2$ s.t.

$\frac{7}{10}x_1 + x_2 \leq 630 $ \
$\frac{1}{2}x_1 + \frac{5}{6}x_2 \leq 600$ \
$x_1 + \frac{2}{3}x_2 \leq 708$ \
$\frac{1}{10}x_1 + \frac{1}{4}x_2 \leq 135$

#### Formulation of the dual problem for the second strategy

max $g(\lambda) = 630\lambda_1 + 600\lambda_2 + 708\lambda_3 + 135\lambda_4$ s.t. \
$\frac{7}{10}\lambda_1 + \frac{1}{2}\lambda_2 + \lambda_3 + \frac{1}{10}\lambda_4 \geq -14$ \
$\lambda_1 + \frac{5}{6}\lambda_2 + \frac{2}{3}\lambda_3 + \frac{1}{4}\lambda_4 \geq -9 $


#### Solution of the dual problem for the second strategy

In [22]:
c = np.array([[630, 600, 708, 135]])
A = np.array([[-7/10, -1/2, -1, -1/10],[-1, -5/6, -2/3, -1/4],[-1,0,0,0],[0,-1,0,0],[0,0,-1,0],[0,0,0,-1]])
b = np.array([[-14],[-9],[0],[0],[0],[0]])

x1_bounds = (0, None)
x2_bounds = (0, None)
x3_bounds = (0, None)
x4_bounds = (0, None)


res = linprog(c, A_ub=A, b_ub=b, bounds = np.array([x1_bounds, x2_bounds, x3_bounds, x4_bounds]), method='simplex')
print(f'The solution to this optimization problem is λ = {res.x}')
print(f'The value of the objective function of the dual evaluated at the optimal value is {round(res.fun,3)}')

The solution to this optimization problem is λ = [ 0.  0. 14.  0.]
The value of the objective function of the dual evaluated at the optimal value is 9912.0


#### Solution of the primal via complementary slackness

From the previous solution, the complementary slackness conditions are evaluated to find the solution in the primal problem, i.e., the quantities of standard and deluxe bags to manufacture. The following constraints are obtained:

$-\frac{7}{10}(0) - \frac{1}{2}(0) - 1(14) -\frac{1}{10}(0) < -14$\
$-14 < -14 \longrightarrow$ Active constraint 

$-1(0) - \frac{5}{6}(0) - \frac{2}{3}(14) -\frac{1}{4}(0) < -9$
$-9.33 = -9 \longrightarrow$ Inactive constraint 

$\lambda_1^*=0\longrightarrow$ Active constraint 

$\lambda_2^*=0\longrightarrow$ Active constraint 

$\lambda_3^*>0\longrightarrow$ Inactive constraint 

$\lambda_4^*=0\longrightarrow$ Active constraint 

From the second and fifth constraints, it is obtained respectively that:

$x_2^*=0$ \
$-x_1^*-\frac{2}{3}x_2^*=-708$

Finally, the system is solved and it is obtained that 708 standard bags and 0 deluxe bags should be manufactured.

### c) 
#### Formulation of the third strategy

Decision variables:

$x_1 \longrightarrow $ Quantity of standard bags \
$x_2 \longrightarrow $ Quantity of deluxe bags

Objective function: Profit

$f(x) = 7x_1 + 9x_2$

Constraints: Time in hours

$\frac{7}{10}x_1 + x_2 \leq 1071 $ \
$\frac{1}{2}x_1 + \frac{5}{6}x_2 \leq 1020$ \
$x_1 + \frac{2}{3}x_2 \leq 1203.6$ \
$\frac{1}{10}x_1 + \frac{1}{4}x_2 \leq 229.5$

Optimization problem: \
max $f(x) = 7x_1 + 9x_2$ s.t.

$\frac{7}{10}x_1 + x_2 \leq 1071 $ \
$\frac{1}{2}x_1 + \frac{5}{6}x_2 \leq 1020$ \
$x_1 + \frac{2}{3}x_2 \leq 1203.6$ \
$\frac{1}{10}x_1 + \frac{1}{4}x_2 \leq 229.5$

#### Formulation of the dual problem for the third strategy

max $g(\lambda) = 1071\lambda_1 + 1020\lambda_2 + 1203.6\lambda_3 + 229.5\lambda_4$ s.t. \
$\frac{7}{10}\lambda_1 + \frac{1}{2}\lambda_2 + \lambda_3 + \frac{1}{10}\lambda_4 \geq -14$ \
$\lambda_1 + \frac{5}{6}\lambda_2 + \frac{2}{3}\lambda_3 + \frac{1}{4}\lambda_4 \geq -9 $


In [23]:
c = np.array([[630*1.7, 600*1.7, 708*1.7, 135*1.7]])
A = np.array([[-7/10, -1/2, -1, -1/10],[-1, -5/6, -2/3, -1/4],[-1,0,0,0],[0,-1,0,0],[0,0,-1,0],[0,0,0,-1]])
b = np.array([[-7],[-9],[0],[0],[0],[0]])

x1_bounds = (0, None)
x2_bounds = (0, None)
x3_bounds = (0, None)
x4_bounds = (0, None)


res = linprog(c, A_ub=A, b_ub=b, bounds = np.array([x1_bounds, x2_bounds, x3_bounds, x4_bounds]), method='simplex')
print(f'The solution to this optimization problem is λ = {res.x}')
print(f'The value of the objective function of the dual evaluated at the optimal value is {round(res.fun,3)}')

The solution to this optimization problem is λ = [8.125  0.     1.3125 0.    ]
The value of the objective function of the dual evaluated at the optimal value is 10281.6


#### Solution of the primal via complementary slackness


From the previous solution, the complementary slackness conditions are evaluated to find the solution in the primal problem, i.e., the quantities of standard and deluxe bags to manufacture. The following constraints are obtained:

$-\frac{7}{10}(8.125) - \frac{1}{2}(0) - 1(1.3125) -\frac{1}{10}(0) < -7$\
$-7 = -7 \longrightarrow$ Active constraint 

$-1(8.125) - \frac{5}{6}(0) - \frac{2}{3}(1.3125) -\frac{1}{4}(0) < -9$
$-9 = -9 \longrightarrow$ Active constraint 

$\lambda_1^*>0\longrightarrow$ Inactive constraint 

$\lambda_2^*=0\longrightarrow$ Active constraint 

$\lambda_3^*>0\longrightarrow$ Inactive constraint 

$\lambda_4^*=0\longrightarrow$ Active constraint 

From the third and fifth constraints, it is obtained respectively that:

$-\frac{7}{10}x_1^* - x_2^*=-1071$ \
$-x_1^*-\frac{2}{3}x_2^*=-1203.6$

Finally, the system is solved and it is obtained that 918 standard bags and 428.4 deluxe bags should be manufactured.


## Strategy Selection

It is more convenient to choose the third strategy because this way a higher total profit value is obtained compared to the other two strategies. Below is the summarized information:

Profit with **first strategy**: $9,720

Profit with **second strategy**: $9,912

Profit with **third strategy**: $10,281.6


# Problem 3

### Different strategies

*PAR*'s store manager has proposed two new strategies to try and maximize profits. 

**Strategy 1:** Training all employees, improving all production times by 10%

**Strategy 2:** A new image that would allow increasing the profit per bag by 10%

Formulate an optimization problem for each strategy and decide which one is the best alternative to maximize profits for the company.

### a) Training all employees, improving all production times by 10%


In [24]:
c = np.array([[630, 600, 708, 135]])
A = np.array([[(-7/10)*0.9, (-1/2)*0.9, -1*0.9, (-1/10)*0.9],[-1*0.9, (-5/6)*0.9, (-2/3)*0.9, (-1/4)*0.9],[-1,0,0,0],[0,-1,0,0],[0,0,-1,0],[0,0,0,-1]])
b = np.array([[-7],[-9],[0],[0],[0],[0]])

x1_bounds = (0, None)
x2_bounds = (0, None)
x3_bounds = (0, None)
x4_bounds = (0, None)


res = linprog(c, A_ub=A, b_ub=b, bounds = np.array([x1_bounds, x2_bounds, x3_bounds, x4_bounds]), method='simplex', options={"disp": True})
print(f'The solution to this optimization problem is λ = {res.x}')
print(f'The value of the objective function of the dual evaluated at the optimal value is {round(res.fun,3)}')

Optimization terminated successfully.
         Current function value: 6720.000000 
         Iterations: 3
The solution to this optimization problem is λ = [9.02777778 0.         1.45833333 0.        ]
The value of the objective function of the dual evaluated at the optimal value is 6720.0


#### Solution of the primal via complementary slackness

From the previous solution, the complementary slackness conditions are evaluated to find the solution in the primal problem, i.e., the quantities of standard and deluxe bags to manufacture. The following constraints are obtained:

$-0.63(9.028) - 0.45(0) - 0.9(1.458) -0.09(0) < -7$
$-7 = -7 \longrightarrow$ Active constraint 

$-0.9(9.028) - 0.75(0) - 0.6(1.458) - 0.225(0) < -9$\
$-9 = -9 \longrightarrow$ Active constraint 

$\lambda_1^*>0\longrightarrow$ Inactive constraint 

$\lambda_2^*=0\longrightarrow$ Active constraint 

$\lambda_3^*>0\longrightarrow$ Inactive constraint 

$\lambda_4^*=0\longrightarrow$ Active constraint 

From the third and fifth constraints, it is obtained respectively that:

$-0.63x_1^* - 0.9x_2^*=-630$ \
$-0.9x_1^*- 0.6x_2^*=-708$

Finally, the system is solved and it is obtained that 600 standard bags and 280 deluxe bags should be manufactured.


In [25]:
Aprim = np.array([[(-7/10)*0.9, -1*0.9],[(-1/2)*0.9,(-5/6)*0.9],[-1*0.9, (-2/3)*0.9],[(-1/10)*0.9,(-1/4)*0.9],[1,0],[0,1]])
bprim = np.array([[-630],[-600],[-708],[-135],[0],[0]])
fprim = np.array([-7, -9])
solution = res.x
solution=solution.reshape(-1,1)
optimal_value = res.fun

sizeAprim = np.shape(Aprim)
sizebprim = np.shape(bprim)
sizefprim = np.shape(fprim)
sizesolution = np.shape(solution)

fdual = bprim[bprim != 0]
sizefdual = np.shape(fdual)

Adual = np.zeros((sizeAprim[1],np.shape(fdual)[0]))
bdual = np.zeros((sizeAprim[1],1))

for i in range(sizeAprim[1]):
    bdual[i] = fprim[i]
    for j in range(sizefdual[0]):
        Aprim_copy = Aprim.copy()
        Aprim_copy = Aprim[:-sizeAprim[1], :]
        Adual[i,j] = Aprim_copy[j][i]
        
sizeAdual = np.shape(Adual)
sizebdual = np.shape(bdual)
sizefdual = np.shape(fdual)

print(f'A from primal: \n {Aprim}')
print(f'b from primal: \n {bprim}')
print(f'f from primal: \n {fprim}')  

print(f'A from dual: \n {Adual}')
print(f'b from dual: \n {bdual}')
print(f'f from dual: \n {fdual}')

sol_primal = np.zeros((sizeAdual[1],sizeAdual[0]))

for i in range(sizeAdual[0]):
    filai = Adual[i]
    value = np.dot(filai,solution)
    if value == bdual[i]:
        pass
    elif value < bdual[i]:
        row = sol_primal[i,:]
        for j in range(np.shape(row)[0]):
            if j == i:
                sol_primal[j+1,i] = 1

bprim_aux = np.zeros((sizesolution[0],1))

for i in range(sizesolution[0]):
    if solution[i] > 0:
        sol_primal[i,:] = Aprim[i,:] 
        bprim_aux[i] = bprim[i]   
    elif solution[i] == 0:
        sol_primal[i,:] = np.zeros(sizeAprim[1])
        bprim_aux[i] = 0

sol_primal = sol_primal[np.all(sol_primal != 0, axis=1)]
bprim_aux = bprim_aux[np.all(bprim_aux != 0, axis=1)]
        
solx = np.linalg.solve(sol_primal, bprim_aux)
        
print(f'Matrix to solve the primal: \n {sol_primal}')
print(f'Vector b to solve the primal: \n {bprim_aux}')
print(f'Values of x: \n {solx}')
print(f'Objective function of the primal: \n {-1*np.dot(fprim,solx)}')


A from primal: 
 [[-0.63  -0.9  ]
 [-0.45  -0.75 ]
 [-0.9   -0.6  ]
 [-0.09  -0.225]
 [ 1.     0.   ]
 [ 0.     1.   ]]
b from primal: 
 [[-630]
 [-600]
 [-708]
 [-135]
 [   0]
 [   0]]
f from primal: 
 [-7 -9]
A from dual: 
 [[-0.63  -0.45  -0.9   -0.09 ]
 [-0.9   -0.75  -0.6   -0.225]]
b from dual: 
 [[-7.]
 [-9.]]
f from dual: 
 [-630 -600 -708 -135]
Matrix to solve the primal: 
 [[-0.63 -0.9 ]
 [-0.9  -0.6 ]]
Vector b to solve the primal: 
 [[-630.]
 [-708.]]
Values of x: 
 [[600.]
 [280.]]
Objective function of the primal: 
 [6720.]


Based on this result, it is obtained that **600 standard bags and 280 deluxe bags should be manufactured** to maximize the profit to a value of **$6,720**.

### b) A new image that would allow increasing the profit per bag by 10%

In [26]:
c = np.array([[630, 600, 708, 135]])
A = np.array([[(-7/10), (-1/2), -1, (-1/10)],[-1, (-5/6), (-2/3), (-1/4)],[-1,0,0,0],[0,-1,0,0],[0,0,-1,0],[0,0,0,-1]])
b = np.array([[-7*1.1],[-9*1.1],[0],[0],[0],[0]])

x1_bounds = (0, None)
x2_bounds = (0, None)
x3_bounds = (0, None)
x4_bounds = (0, None)


res = linprog(c, A_ub=A, b_ub=b, bounds = np.array([x1_bounds, x2_bounds, x3_bounds, x4_bounds]), method='simplex', options={"disp": True})
print(f'The solution to this optimization problem is λ = {res.x}')
print(f'The value of the objective function of the dual evaluated at the optimal value is {round(res.fun,3)}')

Optimization terminated successfully.
         Current function value: 6652.800000 
         Iterations: 3
The solution to this optimization problem is λ = [8.9375  0.      1.44375 0.     ]
The value of the objective function of the dual evaluated at the optimal value is 6652.8


#### Solution of the primal through complementary slackness

From the previous solution, the complementary slackness conditions are evaluated to find the solution in the primal problem, i.e., the quantities of standard and deluxe bags to manufacture. The following constraints are obtained: 

$-\frac{7}{10}(8.9375) - \frac{1}{2}(0) - 1(1.44375) -\frac{1}{10}(0) < -7.7$
$-7.7 = -7.7 \longrightarrow$ Active constraint 

$-1(8.9375) - \frac{5}{6}(0) - \frac{2}{3}(1.44375) -\frac{1}{4}(0) < -9.9$
$-9.9 = -9.9 \longrightarrow$ Active constraint 

$\lambda_1^*>0\longrightarrow$ Inactive constraint 

$\lambda_2^*=0\longrightarrow$ Active constraint 

$\lambda_3^*>0\longrightarrow$ Inactive constraint 

$\lambda_4^*=0\longrightarrow$ Active constraint 

From the third and fifth constraints, it is obtained respectively that:

$-\frac{7}{10}x_1^* - x_2^*=-630$ \
$-x_1^*-\frac{2}{3}x_2^*=-708$

Finally, the system is solved and it is obtained that 540 standard bags and 252 deluxe bags should be manufactured.


In [27]:
Aprim = np.array([[(-7/10), -1],[(-1/2),(-5/6)],[-1, (-2/3)],[(-1/10),(-1/4)],[1,0],[0,1]])
bprim = np.array([[-630],[-600],[-708],[-135],[0],[0]])
fprim = np.array([-7*1.1, -9*1.1])

solucion = res.x

sizeAprim = np.shape(Aprim)
sizebprim = np.shape(bprim)
sizefprim = np.shape(fprim)
sizesolucion = np.shape(solucion)

fdual = bprim[bprim != 0]
sizefdual = np.shape(fdual)

Adual = np.zeros((sizeAprim[1],np.shape(fdual)[0]))
bdual = np.zeros((sizeAprim[1],1))

for i in range(sizeAprim[1]):
    bdual[i] = fprim[i]
    for j in range(sizefdual[0]):
        Aprim_copy = Aprim.copy()
        Aprim_copy = Aprim[:-sizeAprim[1], :]
        Adual[i,j] = Aprim_copy[j][i]
        
sizeAdual = np.shape(Adual)
sizebdual = np.shape(bdual)
sizefdual = np.shape(fdual)

print(f'A from primal: \n {Aprim}')
print(f'b from primal: \n {bprim}')
print(f'f from primal: \n {fprim}')  

print(f'A from dual: \n {Adual}')
print(f'b from dual: \n {bdual}')
print(f'f from dual: \n {fdual}')

sol_primal = np.zeros((sizeAdual[1],sizeAdual[0]))

for i in range(sizeAdual[0]):
    filai = Adual[i]
    valor = np.dot(filai,solucion)
    if valor == bdual[i]:
        pass
    elif valor < bdual[i]:
        fila = sol_primal[i,:]
        for j in range(np.shape(fila)[0]):
            if j == i:
                sol_primal[j+1,i] = 1

bprim_aux = np.zeros((sizesolucion[0],1))

for i in range(sizesolucion[0]):
    if solucion[i] > 0:
        sol_primal[i,:] = Aprim[i,:] 
        bprim_aux[i] = bprim[i]   
    elif solucion[i] == 0:
        sol_primal[i,:] = np.zeros(sizeAprim[1])
        bprim_aux[i] = 0

sol_primal = sol_primal[np.all(sol_primal != 0, axis=1)]
bprim_aux = bprim_aux[np.all(bprim_aux != 0, axis=1)]
        
solx = np.linalg.solve(sol_primal, bprim_aux)
        
print(f'Matrix to solve the primal: \n {sol_primal}')
print(f'Vector b to solve the primal: \n {bprim_aux}')
print(f'Values of x: \n {solx}')
print(f'Objective function of the primal: \n {-1*np.dot(fprim,solx)}')


A from primal: 
 [[-0.7        -1.        ]
 [-0.5        -0.83333333]
 [-1.         -0.66666667]
 [-0.1        -0.25      ]
 [ 1.          0.        ]
 [ 0.          1.        ]]
b from primal: 
 [[-630]
 [-600]
 [-708]
 [-135]
 [   0]
 [   0]]
f from primal: 
 [-7.7 -9.9]
A from dual: 
 [[-0.7        -0.5        -1.         -0.1       ]
 [-1.         -0.83333333 -0.66666667 -0.25      ]]
b from dual: 
 [[-7.7]
 [-9.9]]
f from dual: 
 [-630 -600 -708 -135]
Matrix to solve the primal: 
 [[-0.7        -1.        ]
 [-1.         -0.66666667]]
Vector b to solve the primal: 
 [[-630.]
 [-708.]]
Values of x: 
 [[540.]
 [252.]]
Objective function of the primal: 
 [6652.8]


According to this result, it is obtained that **540 standard bags and 252 deluxe bags** should be manufactured to maximize the profit at a value of **$6,652.8**.


## Strategy selection

Of these last two strategies, it is more convenient to implement the **first** strategy because this way a higher profit value is obtained.

Profit with first strategy: **$6,720**

Profit with second strategy: **$6,652.8**


# Problem 4

A function is defined that returns the primal solution from the dual information, using complementary slackness conditions.

In [28]:
def hcomp(Aprim, bprim, fprim, dual_solution):
    
    # Decompose the dual solution parameter to determine the optimal solution value and the value of the objective function evaluated at that optimal solution.
    solution = dual_solution[0]
    optimal_value = dual_solution[1]
    
    fprim = -1 * fprim
    
    # Obtain the dual system.
    Adual = Aprim.transpose()
    fdual = bprim.transpose()
    bdual = fprim.transpose()
    
    # Create lists to save the indices of variables that are set to zero.
    equations = []
    zeros = []
    
    # First condition of hc: Replace the solution value in the dual constraints. The result is saved in condition11.
    condition11 = Adual.dot(solution).transpose()
    
    # Determine if the constraints of the first hc condition are active or inactive. 
    # False -> Active, True -> Inactive
    constraints = [condition11[k] > bdual[k] for k in range(len(condition11))]
    
    for j in range(len(constraints)):
        # If constraint j is inactive, it means that variable xj = 0. Add the index to a list.
        if constraints[j]:
            zeros.append(j)
    
    # Second condition of hc: If the constraints are inactive, take the index of the equation that will be part of the solution.
    for i in range(len(solution)):
        if solution[i] != 0:
            equations.append(i)
    
    # Create the matrix to solve the system of equations resulting from active and inactive constraints,
    # as well as its corresponding vector b.
    solution_matrix = Aprim.copy()[equations, :]
    b_solution = bprim.copy()[equations, :] 
    
    # Delete the columns from the system and the objective function where xj is zero.
    # Concatenate the resulting system with the b vector (augmented matrix).
    solution_matrix = np.delete(solution_matrix, zeros, axis=1)
    fprim = np.delete(fprim, zeros, axis=1)
    augmented_matrix = np.c_[solution_matrix, b_solution]
    
    # Take the indices of the pivot columns to know the indices of the equations being taken.
    reduced_matrix, indices = sympy.Matrix(augmented_matrix).rref()
    indices = np.array(indices)
    
    # Redefine the matrix only with the relevant equations, as well as the vector b.
    solution_matrix = solution_matrix[indices]
    b_solution = b_solution[indices]
    
    # Find the solution if there are the same number of resulting equations as values in the b vector.
    if len(indices) == len(b_solution):
        x = np.linalg.inv(solution_matrix) @ b_solution

    # Determine the value of the primal function evaluated at the recently found solution.
    primal_value = np.dot(fprim, x)
    
    # Insert the null variables back into their corresponding position to indicate which ones took a value equal to zero.
    x = np.insert(x, zeros, 0)
    
    return primal_value, x


### Function implemented with problem 2 a):

In [30]:
# Constraint matrix
An = np.array([[7/10, 1], [1/2, 5/6], [1, 2/3], [1/10, 1/4]])

# Upper bounds of constraints
b = np.array([[630], [600], [708], [135]])

# Objective function
f = np.array([[-7, -18]])

# Dual solution. The first row corresponds to the solution of the dual problem. The second row corresponds to the value
# of the solution evaluated in the objective function.
sol = [[0, 0, 0, 72], [9720]]

# Apply the function
[f_prim, x] = hcomp(An, b, f, sol)

print(f'The solution to the original optimization problem is x = {x}')
print(f'The value of the objective function evaluated at the optimal value is {f_prim}')

The solution to the original optimization problem is x = [  0. 540.]
The value of the objective function evaluated at the optimal value is [[9720.]]


### Function implemented with problem 2 b):


In [31]:
# Constraint matrix
A = np.array([[7/10, 1], [1/2, 5/6], [1, 2/3], [1/10, 1/4]])

# Upper bounds of constraints
b = np.array([[630], [600], [708], [135]])

# Objective function
f = np.array([[-14, -7]])

# Dual solution. The first row corresponds to the solution of the dual problem. The second row corresponds to the value
# of the solution evaluated in the objective function.
sol = [[0, 0, 14, 0], [9912]]

# Apply the function
[f_prim, x] = hcomp(A, b, f, sol)

print(f'The solution to the original optimization problem is x = {x}')
print(f'The value of the objective function evaluated at the optimal value is {f_prim}')


The solution to the original optimization problem is x = [708.   0.]
The value of the objective function evaluated at the optimal value is [[9912.]]


### Function implemented with problem 2 c):


In [32]:
# Constraint matrix
A = np.array([[7/10, 1], [1/2, 5/6], [1, 2/3], [1/10, 1/4]])

# Upper bounds of constraints
b = np.array([[1071], [1020], [1203.6], [229.5]])

# Objective function
f = np.array([[-7, -9]])

# Dual solution. The first row corresponds to the solution of the dual problem. The second row corresponds to the value
# of the solution evaluated in the objective function.
sol = [[8.125, 0, 1.3125, 0], [10281.6]]

# Apply the function
[f_prim, x] = hcomp(A, b, f, sol)

print(f'The solution to the original optimization problem is x = {x}')
print(f'The value of the objective function evaluated at the optimal value is {f_prim}')

The solution to the original optimization problem is x = [918.  428.4]
The value of the objective function evaluated at the optimal value is [[10281.6]]
